integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/columnLineage/ColumnLevelLineageUtils.java

```diff
@@ -0,0 +1,69 @@
+package io.openlineage.spark3.agent.lifecycle.plan.columnLineage;
+
+import io.openlineage.client.OpenLineage;
+import io.openlineage.spark.api.OpenLineageContext;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * Utility functions for detecting column level lineage within {@link
+ * org.apache.spark.sql.catalyst.plans.logical.LogicalPlan}.
+ */
+@Slf4j
+public class ColumnLevelLineageUtils {
+
+  public static Optional<OpenLineage.ColumnLineageDatasetFacet> buildColumnLineageDatasetFacet(
+      OpenLineageContext context, StructType outputSchema) {
+    if (!context.getQueryExecution().isPresent()
+        || context.getQueryExecution().get().optimizedPlan() == null) {
+      return Optional.empty();
+    }
+
+    ColumnLevelLineageBuilder builder = new ColumnLevelLineageBuilder(outputSchema, context);
+    LogicalPlan plan = context.getQueryExecution().get().optimizedPlan();
+
+    new ExpressionDependencyCollector(plan).collect(builder);
+    new OutputFieldsCollector(plan).collect(builder);
+    new InputFieldsCollector(plan, context).collect(builder);
+
+    OpenLineage.ColumnLineageDatasetFacetBuilder facetBuilder =
+        context.getOpenLineage().newColumnLineageDatasetFacetBuilder();
+
+    facetBuilder.fields(builder.build());
+    OpenLineage.ColumnLineageDatasetFacet facet = facetBuilder.build();
+
+    if (facet.getFields().getAdditionalProperties().isEmpty()) {
+      return Optional.empty();
+    } else {
+      return Optional.of(facetBuilder.build());
+    }
+  }
+
+  public static OpenLineage.OutputDataset rewriteOutputDataset(
+      OpenLineage.OutputDataset outputDataset,
```

Comment:
If you look at usages of this method, it get's applied after `OutputDataset` is already created.  Could you elaborate more on `AbstractOutputDatasetFacetBuilder`? I did not get an idea. 

I think the main problem is on the OL java client side within `OpenLineage.OutputDatasetBuilder` which should also accept builders for the `facets` fields. This would end up with builder of builders and calling a single `build` method should build all the fields. Solution to that exceeds this PR so I decided to move on with workaround like this.